crates/aquascope/src/analysis/permissions/flow.rs

```diff
@@ -0,0 +1,501 @@
+        for local in self.contains_local.iter(from) {
+          if self.dangling_local_sources.contains(local) {
+            return FlowEdgeKind::LocalInvalidatedAtExit;
+          }
+        }
+
+        FlowEdgeKind::Ok
+      }
+    }
+  }
+}
+
+pub fn compute_flows(ctxt: &mut PermissionsCtxt) {
+  let tcx = ctxt.tcx;
+  let body = &ctxt.body_with_facts.body;
+
+  macro_rules! flat_nodes {
+    ($tups:expr) => {
+      $tups
+        .iter()
+        .flat_map(|&(o1, o2)| SmallVec::<[Origin; 2]>::from_buf([o1, o2]))
+        .unique()
+    };
+  }
+
+  macro_rules! node_count {
+    ($tups:expr) => {
+      flat_nodes!($tups)
+        .minmax_by_key(|v| v.index())
+        .into_option()
+        .map(|(_, mx)| mx.index() + 1)
+        .unwrap_or(0)
+    };
+  }
+
+  // Compute the constraint graph with all regions.
+  let constraints = ctxt
+    .polonius_input_facts
+    .subset_base
+    .iter()
+    .map(|&(o1, o2, _)| (o1, o2))
+    .collect::<Vec<_>>();
+
+  let vertices = flat_nodes!(constraints).collect::<Vec<_>>();
+
+  // NOTE: regions that only occur in the return type are not
+  // included in the abstract placeholders set. Example:
+  // ```rust,no-run
+  // fn mk_string() -> &'a String {
+  //   let s = String::from("s");
+  //   &s
+  // }
+  // ```
+  // The `'a`, included for clarity but would be implicit usually,
+  // does not appear in the placeholders set.
+  let placeholders = ctxt
+    .polonius_input_facts
+    .placeholder
+    .iter()
+    .filter_map(|&(p, _)| vertices.contains(&p).then_some(p))
+    .chain(body.regions_in_return().map(|rg| rg.to_region_vid()))
+    .collect::<Vec<_>>();
+
+  log::debug!("PLACEHOLDERS: {placeholders:#?}");
+
+  let placeholder_edges = ctxt
+    .polonius_input_facts
+    .known_placeholder_subset
+    .iter()
+    .filter(|(f, t)| vertices.contains(f) && vertices.contains(t))
+    .copied()
+    .chain(placeholders.iter().map(|&o| (o, o)))
+    .collect::<Vec<_>>();
+
+  // Graph of constraints that need to be satisfied. This shows
+  // us how data flows from one region into another.
+  let constraint_graph = VecGraph::new(node_count!(constraints), constraints);
+
+  // Allowed flows between abstract regions.
+  let known_flows_graph =
+    VecGraph::new(node_count!(placeholder_edges), placeholder_edges);
+
+  let scc_constraints = Sccs::<Origin, SccIdx>::new(&constraint_graph);
+
+  // Compute the total flow facts (expensive).
+  let num_sccs = scc_constraints.num_sccs();
+  let mut known_flows = BitMatrix::new(num_sccs, num_sccs);
+
+  for &r1 in placeholders.iter() {
+    for r2 in known_flows_graph.depth_first_search(r1) {
+      let s1 = scc_constraints.scc(r1);
+      let s2 = scc_constraints.scc(r2);
+      known_flows.insert(s1, s2);
+    }
+  }
+
+  let mut local_sources = BitSet::new_empty(num_sccs);
+  for (_, bd) in ctxt.borrow_set.location_map.iter() {
+    if !(bd.borrowed_place.is_indirect()
+      || bd.borrowed_place.ty(body, tcx).ty.is_ref())
```

Comment:
Suspicious of this condition. What about:

```rust
struct Foo<'a>(&'a i32);
fn f(f: Foo<'_>) {
  let x = &f;
}
```

I believe `x` would be a local source under this rule, but I'm guessing it shouldn't be?